src/Lecture2.hs

```diff
@@ -175,64 +239,80 @@ challenging. However, you still may find some of them easier than some of the
+-- sorted).
+--
+-- >>> isIncreasing [3, 1, 2]
+-- False
+-- >>> isIncreasing [1 .. 10]
+-- True
 isIncreasing :: [Int] -> Bool
-isIncreasing = error "TODO"
-
-{- | Implement a function that takes two lists, sorted in the
-increasing order, and merges them into new list, also sorted in the
-increasing order.
-
-The lists are guaranteed to be given sorted, so you don't need to
-verify that.
-
->>> merge [1, 2, 4] [3, 7]
-[1,2,3,4,7]
--}
+isIncreasing =
+  iterCheck Nothing
+  where
+    iterCheck _ [] = True
+    iterCheck Nothing (h : t) = iterCheck (Just h) t
+    iterCheck (Just a) (h : t)
+      | h < a = False
+      | otherwise = iterCheck (Just h) t
+
+-- | Implement a function that takes two lists, sorted in the
+-- increasing order, and merges them into new list, also sorted in the
+-- increasing order.
+--
+-- The lists are guaranteed to be given sorted, so you don't need to
+-- verify that.
+--
+-- >>> merge [1, 2, 4] [3, 7]
+-- [1,2,3,4,7]
 merge :: [Int] -> [Int] -> [Int]
-merge = error "TODO"
-
-{- | Implement the "Merge Sort" algorithm in Haskell. The @mergeSort@
-function takes a list of numbers and returns a new list containing the
-same numbers but in the increasing order.
-
-The algorithm of merge sort is the following:
-
-  1. If the given list has less than 2 elements, it's already sorted.
-  2. Otherwise, split list into two lists of the same size.
-  3. Sort each of two lists recursively.
-  4. Merge two resulting sorted lists to get a new sorted list.
-
->>> mergeSort [3, 1, 2]
-[1,2,3]
--}
+merge =
+  iter
+  where
+    iter [] r = r
+    iter l [] = l
+    iter (hl : tl) (hr : tr)
+      | hl >= hr = hr : iter (hl : tl) tr
+      | otherwise = hl : iter tl (hr : tr)
+
+-- | Implement the "Merge Sort" algorithm in Haskell. The @mergeSort@
+-- function takes a list of numbers and returns a new list containing the
+-- same numbers but in the increasing order.
+--
+-- The algorithm of merge sort is the following:
+--
+--  1. If the given list has less than 2 elements, it's already sorted.
+--  2. Otherwise, split list into two lists of the same size.
+--  3. Sort each of two lists recursively.
+--  4. Merge two resulting sorted lists to get a new sorted list.
+--
+-- >>> mergeSort [3, 1, 2]
+-- [1,2,3]
 mergeSort :: [Int] -> [Int]
-mergeSort = error "TODO"
-
-
-{- | Haskell is famous for being a superb language for implementing
-compilers and interpreters to other programming languages. In the next
-tasks, you need to implement a tiny part of a compiler.
-
-We're going to work on a small subset of arithmetic operations.
-
-In programming we write expressions like "x + 1" or "y + x + 10".
-Such expressions can be represented in a more structured way (than a
-string) using the following recursive Algebraic Data Type:
--}
+mergeSort [] = []
+mergeSort [a] = [a]
+mergeSort l =
+  merge (mergeSort (halfl l)) (mergeSort (halfr l))
+  where
+    n = length l
+    halfl l = take (div n 2) l
```

Comment:
You can use `splitAt` instead of `take + drop` to avoid traversing the first half twice.

Additionally, you can avoid using `length` if you notice that you don't necessarily need to split in the middle. You only need to split it into two lists of approximately the same size.

In that case, you can implement a solution that traverses a list only once and splits its elements into two lists of elements on only even positions and only odd positions.

Alternatively, you can implement a bottom-up approach by converting a list of elements to a list of of singleton lists and merging them pairwise until only one list is left.